IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -0,0 +1,220 @@
+        self.queue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
+- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects setObject:credential forKey:key];
+    });
+}
+
+- (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem
+{
+    NSArray *keys = [storageItem allKeys];
+    for (NSString *key in keys)
+    {
+        if (![self storedCredentialForKey:key])
+        {
+            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
+            if (credential)
+            {
+                [self storeCredential:credential forKey:key];
+            }
+        }
+    }
+}
+
+- (MSIDCredentialCacheItem *)storedCredentialForKey:(NSString *)key
+{
+    __block MSIDCredentialCacheItem *credential = nil;
+    
+    dispatch_sync(self.queue, ^{
+        credential = [self.cacheObjects objectForKey:key];
+    });
+    
+    return credential;
+}
+
+- (void)removeStoredCredentialForKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects removeObjectForKey:key];
+    });
+}
+
+- (NSArray<MSIDCredentialCacheItem *> *)storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    if (key.account && key.service)
+    {
+        NSString *credentialKey = [NSString stringWithFormat:@"%@%@%@", key.account, keyDelimiter, key.service];
+        MSIDCredentialCacheItem *credential = [self storedCredentialForKey:credentialKey];
+        if (credential)
+        {
+            return @[credential];
+        }
+        
+        return nil;
+    }
+    
+    NSMutableArray *subPredicates = [[NSMutableArray alloc] init];
+    
+    /*
+     Refresh token either has family Id as 1 or nil.
+     For app refresh token, we pass family id as nil which matches both FRT and RT instead of matching just RT
+     For clients that are not part of family, family is saved as empty string and added to refresh token query.
+     */
+    if ([NSString msidIsStringNilOrBlank:key.familyId])
+    {
+        key.familyId = nil;
+    }
+
+    [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.familyId == %@", key.familyId]];
+    if (key.clientId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.clientId == %@", key.clientId]];
+    if (key.environment)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.environment == %@", key.environment]];
+    if (key.homeAccountId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.homeAccountId == %@", key.homeAccountId]];
+    if (key.credentialType)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.credentialType == %d", key.credentialType]];
+    if (key.realm)
```

Comment:
what about target?

Reply:
As discussed. Added a TODO for target.:)